File: yep/src/main/java/catchla/yep/view/StaticMapView.java
package catchla.yep.view;

import android.annotation.TargetApi;
import android.content.Context;
import android.location.Location;
import android.os.Build;
import android.util.AttributeSet;
import android.widget.ImageView;

import javax.inject.Inject;

import catchla.yep.util.ImageLoaderWrapper;
import catchla.yep.util.StaticMapUrlGenerator;
import catchla.yep.util.dagger.ApplicationModule;
import catchla.yep.util.dagger.DaggerGeneralComponent;
import catchla.yep.util.dagger.GeneralComponentHelper;

/**
 * Created by mariotaku on 15/12/9.
 */
public class StaticMapView extends ImageView {
    StaticMapUrlGenerator mGenerator = new StaticMapUrlGenerator();
    @Inject
    ImageLoaderWrapper mImageLoader;
    private Location mLocation;
    private int mZoomLevel;
    private boolean mScaleToDensity;

    public StaticMapView(final Context context) {
        super(context);
        init(context);
    }

    public StaticMapView(final Context context, final AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public StaticMapView(final Context context, final AttributeSet attrs, final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public StaticMapView(final Context context, final AttributeSet attrs, final int defStyleAttr, final int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context);
    }

    private void init(Context context) {
        GeneralComponentHelper.build(context).inject(this);
        setScaleType(ScaleType.CENTER_CROP);
    }

    public void setProvider(final StaticMapUrlGenerator.Provider provider) {
        mGenerator.setProvider(provider);
    }

    public void setScaleToDensity(boolean scaleToDensity) {
        mGenerator.setScale(scaleToDensity ? getResources().getDisplayMetrics().density : 1);
        loadImage();
    }

    public void display(Location location, int zoomLevel) {
        mLocation = location;
        mZoomLevel = zoomLevel;
        if (!mGenerator.hasSize()) {
            mGenerator.setSize(getWidth(), getHeight());
        }
        loadImage();
    }

    private void loadImage() {
        if (mLocation == null || !mGenerator.hasSize()) return;
        mImageLoader.displayImage(mGenerator.generate(mLocation, mZoomLevel), this);
    }

    @Override
    protected void onLayout(final boolean changed, final int left, final int top, final int right, final int bottom) {
        super.onLayout(changed, left, top, right, bottom);
    }

    @Override
    protected void onSizeChanged(final int w, final int h, final int oldw, final int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        if (w != oldw || h != oldh) {
            mGenerator.setSize(w, h);
            loadImage();
        }
    }
}

File: src/main/java/catchla/yep/view/StaticMapView.kt
package catchla.yep.view

import android.content.Context
import android.location.Location
import android.util.AttributeSet
import android.widget.ImageView
import catchla.yep.util.ImageLoaderWrapper
import catchla.yep.util.StaticMapUrlGenerator
import catchla.yep.util.dagger.GeneralComponentHelper
import javax.inject.Inject

/**
 * Created by mariotaku on 15/12/9.
 */
class StaticMapView : ImageView {
    internal var generator = StaticMapUrlGenerator()
    @Inject
    lateinit var imageLoader: ImageLoaderWrapper
    private var location: Location? = null
    private var zoomLevel: Int = 0
    private val scaleToDensity: Boolean = false

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    init {
        if (!isInEditMode) {
            GeneralComponentHelper.build(context).inject(this)
        }
        scaleType = ImageView.ScaleType.CENTER_CROP
    }

    fun setProvider(provider: StaticMapUrlGenerator.Provider) {
        generator.setProvider(provider)
    }

    fun setScaleToDensity(scaleToDensity: Boolean) {
        generator.setScale(if (scaleToDensity) resources.displayMetrics.density else 1f)
        loadImage()
    }

    fun display(location: Location, zoomLevel: Int) {
        this.location = location
        this.zoomLevel = zoomLevel
        if (!generator.hasSize()) {
            generator.setSize(width, height)
        }
        loadImage()
    }

    private fun loadImage() {
        if (location == null || !generator.hasSize() || isInEditMode) return
        imageLoader.displayImage(generator.generate(location, zoomLevel), this)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        if (w != oldw || h != oldh) {
            generator.setSize(w, h)
            loadImage()
        }
    }
}
